### ISSUE TITLE
Auto Sync stopped

### ISSUE THREAD

**Issue by jrieken** [microsoft/vscode#96374]
#96226

found this 👇 in my logs, not sure what happened but this might be interesting

```
[2020-04-28 12:35:39.806] [userDataSync] [info] Auto Sync: stopped.
[2020-04-28 12:35:39.807] [userDataSync] [error] Unauthorized (UserDataSyncError) undefined: Request 'https://vscode-sync.trafficmanager.net/v1/manifest' failed because of Unauthorized (401).
    at e.request (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:502:689)
    at async e.manifest (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:501:157)
    at async e.sync (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:493:321)
    at async e.sync (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:480:734)
[2020-04-28 12:35:41.793] [userDataSync] [info] Auto Sync: Started
```

**Comment by sandy081** [comment]
This happens when Sync is turned off or Auth token is not available. 

@jrieken May I know if you turned off sync at that point and turned on again?

**Comment by jrieken** [comment]
no - i have not

**Comment by RMacfarlane** [comment]
Since syncing happens in the shared process, we have a push model for the token. Instead of the shared process requesting an access token each time it needs to make a request, windows write the token the the shared process. (The shared process can't just send out a request to all windows to get the token - instances of the extension are sometimes in different states. For example, when I log in to one window, other windows poll in the background and see that the keychain data has changed, so there is a bit of lag time before they know about the token. Similarly if I log out in one window it takes time for that to propagate to other windows)

Because we are always pushing the token to the share process, the auth provider extension is responsible for refreshing the token before it expires and sending it again. I set a timeout to do this, the expiry is I think around an hour. Setting a timeout over that period of time is not completely reliable - if your computer goes to sleep, the timer drifts that amount of time. So, I'm not always able to refresh it in time. (I remember I looked into OS eventing on wake up, thinking I could check if the token had expired then, and this didn't exist on some or possibly all platforms.)

If the refresh does fail, the shared process gets a 401, and sends an event to all windows that the token failed. All instances of the extension get this and try to refresh the token, if one of them succeeds, the new, refreshed token is pushed to the shared process again.

Anyway, this means that I think the above is expected behavior - the token expired, but sync was able to recover a couple of seconds later
